404 logic change

It seems like https://github.com/azu/can-npm-publish/commit/7967d92 changed the way 404s are treated.

For node 12 and 1.3.4 the exist code is 1 and `npm ERR! code E404` is printed to the terminal
For node 14 and 1.3.4 the exist code is 1 and `npm ERR! code E404` is printed to the terminal
For node 16 and 1.3.4 the exist code is 1 and `npm ERR! code E404` is printed to the terminal

For node 12 and 1.3.3 things are all good
For node 14 and 1.3.3 things are all good
For node 16 and 1.3.3 `can-npm-publish/lib/can-npm-publish.js:68:37` throws an `Unexpected end of JSON input`/`undefined` error

Here is my output for node 16 and 1.3.4:
```
> nvm use 16 
Now using node v16.4.0 (npm v7.18.1)
> npx can-npm-publish@1.3.4 --verbose
npm ERR! code E404
npm ERR! 404 Not Found - GET https://mycompany.com/api/npm/npm/@scoped%2fpkgname- not_found
npm ERR! 404 
npm ERR! 404  '@scoped/pkgname@latest' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.
{
  "error": {
    "code": "E404",
    "summary": "Not Found - GET https://mycompany.com/api/npm/npm/@scoped%2fpkgname- not_found",
    "detail": "\n '@scoped/pkgname@latest' is not in the npm registry.\nYou should bug the author to publish it (or use the name yourself!)\n\nNote that you can also install from a\ntarball, folder, http url, or git url."
  }
}

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/me/.npm/_logs/2021-06-23T17_27_52_658Z-debug.log

> echo $status
1
```

npm 7 has a bug…

npm 7

```
❯ npm -v
7.18.1

~/.ghq/github.com/azu/can-npm-publish master*
❯ npm view not_found_package_123 --json 2>stderr.log 1>stdout.log

~/.ghq/github.com/azu/can-npm-publish master*
❯ cat stdout.log

~/.ghq/github.com/azu/can-npm-publish master*
❯ cat stderr.log
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/not_found_package_123 - Not found
npm ERR! 404
npm ERR! 404  'not_found_package_123@latest' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.
{
  "error": {
    "code": "E404",
    "summary": "Not Found - GET https://registry.npmjs.org/not_found_package_123 - Not found",
    "detail": "\n 'not_found_package_123@latest' is not in the npm registry.\nYou should bug the author to publish it (or use the name yourself!)\n\nNote that you can also install from a\ntarball, folder, http url, or git url."
  }
}

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/azu/.npm/_logs/2021-06-23T23_35_58_106Z-debug.log
```

npm 6

```
~/.ghq/github.com/azu/can-npm-publish master*
❯ npm -v
6.14.13

~/.ghq/github.com/azu/can-npm-publish master*
❯ npm view not_found_package_123 --json 2>stderr.log 1>stdout.log

~/.ghq/github.com/azu/can-npm-publish master*
❯ cat stdout.log
{
  "error": {
    "code": "E404",
    "summary": "'not_found_package_123' is not in the npm registry.\nYou should bug the author to publish it\n(or use the name yourself!)\n\nNote that you can also install from a\ntarball, folder, http url, or git url.",
    "detail": "\n 'not_found_package_123@latest' is not in the npm registry.\nYou should bug the author to publish it (or use the name yourself!)\n\nNote that you can also install from a\ntarball, folder, http url, or git url."
  }
}

~/.ghq/github.com/azu/can-npm-publish master*
❯ cat stderr.log
npm ERR! code E404
npm ERR! 404 'not_found_package_123' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it
npm ERR! 404 (or use the name yourself!)
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.
npm ERR! 404
npm ERR! 404  'not_found_package_123@latest' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/azu/.npm/_logs/2021-06-23T23_36_48_205Z-debug.log
```

--- a/lib/can-npm-publish.js
+++ b/lib/can-npm-publish.js
@@ -75,17 +75,21 @@ const viewPackage = (packageName, registry) => {
         const getJsonOutputs = ({ stdout, stderr }) => {
             let stdoutJSON = {};
             let stderrJSON = {};
-            try {
-                stdoutJSON = JSON.parse(stdout);
-            } catch (error) {
-                // nope
-                // console.warn("stdoutJSON parse error", stdout);
+            if (stdout) {
+                try {
+                    stdoutJSON = JSON.parse(stdout);
+                } catch (error) {
+                    // nope
+                    // console.warn("stdoutJSON parse error", stdout);
+                }
             }
-            try {
-                stderrJSON = JSON.parse(stderr);
-            } catch (error) {
-                // nope
-                // console.warn("stderrJSON parse error", stderr);
+            if (stderr) {
+                try {
+                    stderrJSON = JSON.parse(stderr);
+                } catch (error) {
+                    // nope
+                    // console.warn("stderrJSON parse error", stderr);
+                }
             }
             return {
                 stdoutJSON,

--- a/lib/can-npm-publish.js
+++ b/lib/can-npm-publish.js
@@ -104,7 +104,7 @@ const viewPackage = (packageName, registry) => {
         const isError = (json) => {
             return json && "error" in json;
         };
-        const is404Json = (json) => {
+        const is404Error = (json) => {
             return isError(json) && json.error.code === "E404";
         };
         view.stdout.on("data", (data) => {
@@ -134,11 +134,11 @@ const viewPackage = (packageName, registry) => {
                 stderr: _stderrResult
             });
             // npm6 view --json output to stdout if the package is 404 → can publish
-            if (is404Json(stdoutJSON)) {
+            if (is404Error(stdoutJSON)) {
                 return resolve([]);
             }
             // npm7 view --json output to stderr if the package is 404 → can publish
-            if (is404Json(stderrJSON)) {
+            if (is404Error(stderrJSON)) {
                 return resolve([]);
             }
             // in other error, can not publish → reject

--- a/lib/can-npm-publish.js
+++ b/lib/can-npm-publish.js
@@ -69,32 +69,82 @@ const viewPackage = (packageName, registry) => {
     return new Promise((resolve, reject) => {
         const registryArgs = registry ? ["--registry", registry] : [];
         const view = spawn("npm", ["view", packageName, "versions", "--json"].concat(registryArgs));
-        let result = "";
-        let errorResult = "";
+        let _stdoutResult = "";
+        let _stderrResult = "";
 
+        const getJsonOutputs = ({ stdout, stderr }) => {
+            let stdoutJSON = {};
+            let stderrJSON = {};
+            try {
+                stdoutJSON = JSON.parse(stdout);
+            } catch (error) {
+                // nope
+                // console.warn("stdoutJSON parse error", stdout);
+            }
+            try {
+                stderrJSON = JSON.parse(stderr);
+            } catch (error) {
+                // nope
+                // console.warn("stderrJSON parse error", stderr);
+            }
+            return {
+                stdoutJSON,
+                stderrJSON
+            };
+        };
+        const isError = (json) => {
+            return json && "error" in json;
+        };
+        const is404Json = (json) => {
+            return isError(json) && json.error.code === "E404";
+        };
         view.stdout.on("data", (data) => {
-            result += data.toString();
+            _stdoutResult += data.toString();
         });
 
         view.stderr.on("data", (err) => {
-            errorResult += err.toString();
+            const stdErrorStr = err.toString();
+            try {
+                // Workaround for npm 7
+                // npm 7 --json option is broken
+                // It aim to remove non json output.
+                // FIXME: However,This logics will break json chunk(chunk may be invalid json)
+                // https://github.com/azu/can-npm-publish/issues/19
+                // https://github.com/npm/cli/issues/2740
+                JSON.parse(stdErrorStr); // parse check
+                _stderrResult += stdErrorStr;
+            } catch (error) {}
         });
 
         view.on("close", (code) => {
+            // Note:
+            // npm 6 output JSON in stdout
+            // npm 7(7.18.1) output JSON in stderr
+            const { stderrJSON, stdoutJSON } = getJsonOutputs({
+                stdout: _stdoutResult,
+                stderr: _stderrResult
+            });
+            // npm6 view --json output to stdout if the package is 404 → can publish
+            if (is404Json(stdoutJSON)) {
+                return resolve([]);
+            }
+            // npm7 view --json output to stderr if the package is 404 → can publish
+            if (is404Json(stderrJSON)) {
+                return resolve([]);
+            }
+            // in other error, can not publish → reject
+            if (isError(stdoutJSON)) {
+                return reject(new Error(_stdoutResult));
+            }
+            if (isError(stderrJSON)) {
+                return reject(new Error(_stderrResult));
+            }
+            // if command is failed by other reasons(no json output), treat it as actual error
             if (code !== 0) {
-                return reject(new Error(errorResult));
+                return reject(new Error(_stderrResult));
             }
-            const resultJSON = JSON.parse(result);
-            if (resultJSON && resultJSON.error) {
-                // the package is not in the npm registry => can publish
-                if (resultJSON.error.code === "E404") {
-                    return resolve([]); // resolve as empty version
-                } else {
-                    // other error => can not publish
-                    return reject(new Error(errorResult));
-                }
-            }
-            resolve(resultJSON);
+            // if success to get, resolve with versions json
+            return resolve(stdoutJSON);
         });
     });
 };

--- a/lib/can-npm-publish.js
+++ b/lib/can-npm-publish.js
@@ -140,8 +140,8 @@ const viewPackage = (packageName, registry, options) => {
             });
             if (options.verbose) {
                 console.log("`npm view` command's exit code:", code);
-                console.log("stdoutJSON", stdoutJSON);
-                console.log("stderrJSON", stderrJSON);
+                console.log("`npm view` stdoutJSON", stdoutJSON);
+                console.log("`npm view` stderrJSON", stderrJSON);
             }
             // npm6 view --json output to stdout if the package is 404 → can publish
             if (is404Error(stdoutJSON)) {

--- a/lib/can-npm-publish.js
+++ b/lib/can-npm-publish.js
@@ -63,9 +63,10 @@ const checkPrivateField = (packagePath) => {
  * or rejects if anything failed
  * @param packageName
  * @param registry
+ * @param {{verbose : boolean}} options
  * @returns {Promise}
  */
-const viewPackage = (packageName, registry) => {
+const viewPackage = (packageName, registry, options) => {
     return new Promise((resolve, reject) => {
         const registryArgs = registry ? ["--registry", registry] : [];
         const view = spawn("npm", ["view", packageName, "versions", "--json"].concat(registryArgs));
@@ -85,7 +86,9 @@ const viewPackage = (packageName, registry) => {
                     stdoutJSON = JSON.parse(stdout);
                 } catch (error) {
                     // nope
-                    // console.warn("stdoutJSON parse error", stdout);
+                    if (options.verbose) {
+                        console.error("stdoutJSON parse error", stdout);
+                    }
                 }
             }
             if (stderr) {
@@ -93,7 +96,9 @@ const viewPackage = (packageName, registry) => {
                     stderrJSON = JSON.parse(stderr);
                 } catch (error) {
                     // nope
-                    // console.warn("stderrJSON parse error", stderr);
+                    if (options.verbose) {
+                        console.error("stderrJSON parse error", stdout);
+                    }
                 }
             }
             return {
@@ -129,10 +134,15 @@ const viewPackage = (packageName, registry) => {
             // Note:
             // npm 6 output JSON in stdout
             // npm 7(7.18.1) output JSON in stderr
-            const { stderrJSON, stdoutJSON } = getJsonOutputs({
+            const { stdoutJSON, stderrJSON } = getJsonOutputs({
                 stdout: _stdoutResult,
                 stderr: _stderrResult
             });
+            if (options.verbose) {
+                console.log("`npm view` command's exit code:", code);
+                console.log("stdoutJSON", stdoutJSON);
+                console.log("stderrJSON", stderrJSON);
+            }
             // npm6 view --json output to stdout if the package is 404 → can publish
             if (is404Error(stdoutJSON)) {
                 return resolve([]);
@@ -162,7 +172,13 @@ const viewPackage = (packageName, registry) => {
     });
 };
 
-const checkAlreadyPublish = (packagePath) => {
+/**
+ *
+ * @param {string} packagePath
+ * @param {{verbose : boolean}} options
+ * @returns {Promise<readPkg.NormalizedPackageJson>}
+ */
+const checkAlreadyPublish = (packagePath, options) => {
     return readPkgWithPath(packagePath).then((pkg) => {
         const name = pkg["name"];
         const version = pkg["version"];
@@ -174,7 +190,7 @@ const checkAlreadyPublish = (packagePath) => {
         if (version === undefined) {
             return Promise.reject(new Error("This package has no `version`."));
         }
-        return viewPackage(name, registry).then((versions) => {
+        return viewPackage(name, registry, options).then((versions) => {
             if (versions.includes(version)) {
                 return Promise.reject(new Error(`${name}@${version} is already published`));
             }
@@ -191,7 +207,7 @@ const checkAlreadyPublish = (packagePath) => {
 const canNpmPublish = (packagePath, options = { verbose: false }) => {
     return Promise.all([
         checkPkgName(packagePath, options),
-        checkAlreadyPublish(packagePath),
+        checkAlreadyPublish(packagePath, options),
         checkPrivateField(packagePath)
     ]);
 };

--- a/lib/can-npm-publish.js
+++ b/lib/can-npm-publish.js
@@ -72,9 +72,14 @@ const viewPackage = (packageName, registry) => {
         let _stdoutResult = "";
         let _stderrResult = "";
 
+        /**
+         * @param stdout
+         * @param stderr
+         * @returns {{stdoutJSON: null | {}, stderrJSON: null | {}}}
+         */
         const getJsonOutputs = ({ stdout, stderr }) => {
-            let stdoutJSON = {};
-            let stderrJSON = {};
+            let stdoutJSON = null;
+            let stderrJSON = null;
             if (stdout) {
                 try {
                     stdoutJSON = JSON.parse(stdout);
@@ -147,8 +152,12 @@ const viewPackage = (packageName, registry) => {
             if (code !== 0) {
                 return reject(new Error(_stderrResult));
             }
-            // if success to get, resolve with versions json
-            return resolve(stdoutJSON);
+            if (stdoutJSON) {
+                // if success to get, resolve with versions json
+                return resolve(stdoutJSON);
+            } else {
+                return reject(_stderrResult);
+            }
         });
     });
 };

--- a/lib/can-npm-publish.js
+++ b/lib/can-npm-publish.js
@@ -4,6 +4,7 @@ const path = require("path");
 const spawn = require("cross-spawn");
 const readPkg = require("read-pkg");
 const validatePkgName = require("validate-npm-package-name");
+const { extractJSONObject } = require("extract-first-json");
 /**
  * @param {string} [filePathOrDirPath]
  * @returns {Promise<readPkg.NormalizedPackageJson>}
@@ -118,16 +119,16 @@ const viewPackage = (packageName, registry, options) => {
 
         view.stderr.on("data", (err) => {
             const stdErrorStr = err.toString();
-            try {
-                // Workaround for npm 7
-                // npm 7 --json option is broken
-                // It aim to remove non json output.
-                // FIXME: However,This logics will break json chunk(chunk may be invalid json)
-                // https://github.com/azu/can-npm-publish/issues/19
-                // https://github.com/npm/cli/issues/2740
-                JSON.parse(stdErrorStr); // parse check
-                _stderrResult += stdErrorStr;
-            } catch (error) {}
+            // Workaround for npm 7
+            // npm 7 --json option is broken
+            // It aim to remove non json output.
+            // FIXME: However,This logics will break json chunk(chunk may be invalid json)
+            // https://github.com/azu/can-npm-publish/issues/19
+            // https://github.com/npm/cli/issues/2740
+            const jsonObject = extractJSONObject(stdErrorStr);
+            if (jsonObject) {
+                _stderrResult = JSON.stringify(jsonObject, null, 4);
+            }
         });
 
         view.on("close", (code) => {

--- a/package.json
+++ b/package.json
@@ -39,6 +39,7 @@
   },
   "dependencies": {
     "cross-spawn": "^7.0.3",
+    "extract-first-json": "^1.0.1",
     "meow": "^9.0.0",
     "read-pkg": "^5.0.0",
     "validate-npm-package-name": "^3.0.0"

--- a/package.json
+++ b/package.json
@@ -58,5 +58,9 @@
     "*.{js,jsx,ts,tsx,css}": [
       "prettier --write"
     ]
+  },
+  "volta": {
+    "node": "16.4.0",
+    "npm": "7.18.1"
   }
 }

--- a/test/can-npm-publish-bin-test.js
+++ b/test/can-npm-publish-bin-test.js
@@ -5,10 +5,6 @@ const path = require("path");
 // Path to the executable script
 const binPath = path.join(__dirname, "../bin/cmd.js");
 
-const shouldNotCalled = () => {
-    throw new Error("SHOULD NOT CALLED");
-};
-
 describe("can-npm-publish bin", () => {
     it("should return 0, it can publish", (done) => {
         const bin = spawn("node", [binPath, path.join(__dirname, "fixtures/not-published-yet.json")]);
@@ -16,6 +12,8 @@ describe("can-npm-publish bin", () => {
         // Finish the test when the executable finishes and returns 0
         bin.on("close", (exit_code) => {
             assert.ok(exit_code === 0);
+        });
+        bin.on("close", () => {
             done();
         });
     });
@@ -30,11 +28,12 @@ describe("can-npm-publish bin", () => {
     });
     it("should send errors to stderr when verbose, it can't publish", (done) => {
         const bin = spawn("node", [binPath, path.join(__dirname, "fixtures/already-published.json"), "--verbose"]);
-
         // Finish the test and stop the executable when it outputs to stderr
         bin.stderr.on("data", (data) => {
             assert.ok(/almin@0.15.2 is already published/.test(data));
             bin.kill();
+        });
+        bin.on("close", () => {
             done();
         });
     });

--- a/test/can-npm-publish-test.js
+++ b/test/can-npm-publish-test.js
@@ -38,7 +38,7 @@ describe("can-npm-publish", () => {
         return canNpmPublish(path.join(__dirname, "fixtures/not-published-yet.json"));
     });
     it("should be resolve, it is not 404 package", () => {
-        return canNpmPublish(path.join(__dirname, "fixtures/404-package.json"));
+        return canNpmPublish(path.join(__dirname, "fixtures/404-package.json"), { verbose: true });
     });
     it("should be resolve, it is not published yet to yarnpkg registry", () => {
         return canNpmPublish(path.join(__dirname, "fixtures/not-published-yet-registry.json"));

--- a/test/can-npm-publish-test.js
+++ b/test/can-npm-publish-test.js
@@ -37,6 +37,9 @@ describe("can-npm-publish", () => {
     it("should be resolve, it is not published yet", () => {
         return canNpmPublish(path.join(__dirname, "fixtures/not-published-yet.json"));
     });
+    it("should be resolve, it is not 404 package", () => {
+        return canNpmPublish(path.join(__dirname, "fixtures/404-package.json"));
+    });
     it("should be resolve, it is not published yet to yarnpkg registry", () => {
         return canNpmPublish(path.join(__dirname, "fixtures/not-published-yet-registry.json"));
     });

--- a/test/fixtures/404-package.json
+++ b/test/fixtures/404-package.json
@@ -1,0 +1,5 @@
+{
+  "private": false,
+  "name": "asasaiqjpjopmsonopajrpqwkmxzoj22",
+  "version": "1.0.0"
+}

